main.go

```diff
@@ -29,15 +38,116 @@ func (cfg Config) validateEnvs() error {
-			return fmt.Errorf("neither password nor app_password is provided")
+	switch isJWTAuthType, isAppleIDAuthType := (cfg.APIKeyPath != "" || cfg.APIIssuer != ""), (cfg.AppPassword != "" || cfg.Password != "" || cfg.ItunesConnectUser != ""); {
+
+	case isAppleIDAuthType == isJWTAuthType:
+
+		return fmt.Errorf("one type of authentication required, either provide Apple ID with password/app password or API key with issuer")
+
+	case isAppleIDAuthType:
+
+		if err := input.ValidateIfNotEmpty(string(cfg.ItunesConnectUser)); err != nil {
+			return fmt.Errorf("no Apple ID provided")
+		}
+		if err := input.ValidateIfNotEmpty(string(cfg.Password)); err != nil {
+			if err := input.ValidateIfNotEmpty(string(cfg.AppPassword)); err != nil {
+				return fmt.Errorf("neither password nor app_password is provided")
+			}
+		}
+
+	case isJWTAuthType:
+
+		if err := input.ValidateIfNotEmpty(string(cfg.APIIssuer)); err != nil {
+			return fmt.Errorf("no API Issuer provided")
 		}
+		if err := input.ValidateIfNotEmpty(string(cfg.APIKeyPath)); err != nil {
+			return fmt.Errorf("no API Key path provided")
+		}
+
 	}
 
 	return nil
 }
 
+// prepares key and returns the ID the tool need to use
+func prepareAPIKey(apiKeyPath string) (string, error) {
+	// input should accept url and filepath
+	// 4 location existing, check the files there first
+	// if there is a file on any of the 4 locations use that
+	// if ther eis no that file found then move it to one of the locations
+	// if the filename is not conventional for the tool then create a new unique one
+
+	// see these in the altool's man page
+	var (
+		keyPaths = []string{filepath.Join(os.Getenv("HOME"), ".appstoreconnect/private_keys"), "./private_keys", filepath.Join(os.Getenv("HOME"), "private_keys"), filepath.Join(os.Getenv("HOME"), ".private_keys")}
+		keyID    = "Bitrise" // as default if no ID found in file name
+	)
+
+	// parse string to url
+	fileURL, err := url.Parse(apiKeyPath)
+	if err != nil {
+		return "", err
+	}
+
+	// get the ID of the key from the file
+	if matches := regexp.MustCompile(`AuthKey_(.+)\.p8`).FindStringSubmatch(filepath.Base(fileURL.Path)); len(matches) == 2 {
+		keyID = matches[1]
+	}
+
+	certName := fmt.Sprintf("AuthKey_%s.p8", keyID)
+
+	// if certName already exists on any of the following locations then return that's ID here
+	for _, path := range keyPaths {
+		exists, err := pathutil.IsPathExists(filepath.Join(path, certName))
+		if err != nil {
+			return "", err
+		}
+		if exists {
+			return keyID, nil
+		}
+	}
+
+	// cert file not found on any of the locations, so copy or download it
+
+	// create cert file on the first location
+	certFile, err := os.Create(filepath.Join(keyPaths[0], certName))
+	if err != nil {
+		return "", err
+	}
+	defer func() {
+		if err := certFile.Close(); err != nil {
+			log.Errorf("Failed to close file, error: %s", err)
+		}
+	}()
+
+	// if file -> download
+	if fileURL.Scheme == "file" {
```

Comment:
i would also create a functions for this:

`copyFile(srcPth string, dest os.File) error`
`downloadFile(srcURL string, dest os.File) error`

This way you do not need any comment, in case of "file" URL scheme, copy, otherwise download

Reply:
the only reason I left is in a single code block to share a lot of common, let me know if you think it still would be better to be seperated